vortex-expr/src/lib.rs

```diff
@@ -47,11 +47,30 @@ pub trait VortexExpr: Debug + Send + Sync + DynEq + Display {
     fn as_any(&self) -> &dyn Any;
 
     /// Compute result of expression on given batch producing a new batch
-    fn evaluate(&self, batch: &ArrayData) -> VortexResult<ArrayData>;
+    ///
+    fn evaluate(&self, batch: &ArrayData) -> VortexResult<ArrayData> {
+        let result = self.unchecked_evaluate(batch)?;
+        debug_assert_eq!(result.dtype(), &self.dtype(batch.dtype())?);
+        Ok(result)
+    }
+
+    /// Compute result of expression on given batch producing a new batch
+    ///
+    /// "Unchecked" means that this function lacks a debug assertion that the returned array matches
+    /// the [VortexExpr::dtype] method. Use instead the [VortexExpr::evaluate] function which
+    /// includes such an assertion.
+    fn unchecked_evaluate(&self, batch: &ArrayData) -> VortexResult<ArrayData>;
 
     fn children(&self) -> Vec<&ExprRef>;
 
     fn replacing_children(self: Arc<Self>, children: Vec<ExprRef>) -> ExprRef;
+
+    /// Compute the type of the array returned by [VortexExpr::evaluate].
+    fn dtype(&self, input_dtype: &DType) -> VortexResult<DType> {
```

Comment:
Also if we're going with the word "scope" to refer to whatever the Identity refers to, then can you name the argument `scope_dtype`